File: src/main/java/co/renil/astro/kundli/service/KundliService.java
package co.renil.astro.kundli.service;

import co.renil.astro.kundli.entity.Kundli;
import co.renil.astro.kundli.repository.KundliRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class KundliService {

    private final KundliRepository kundliRepository;

    // Create a new Kundli
    public Kundli createKundli(Kundli kundli) {
        return kundliRepository.save(kundli);
    }

    // Retrieve Kundli by ID
    public Optional<Kundli> getKundliById(UUID kundliId) {
        return kundliRepository.findById(kundliId);
    }

    // Update an existing Kundli
    public Kundli updateKundli(UUID kundliId, Kundli kundliDetails) {
        Optional<Kundli> existingKundli = kundliRepository.findById(kundliId);
        if (existingKundli.isPresent()) {
            Kundli kundli = existingKundli.get();
            // Update properties as needed
            kundli.setBhavaPredictions(kundliDetails.getBhavaPredictions());
            kundli.setNakshatra(kundliDetails.getNakshatra());
            kundli.setLatitude(kundliDetails.getLatitude());
            // Continue updating other fields...
            return kundliRepository.save(kundli);
        }
        throw new RuntimeException("Kundli not found with id " + kundliId);
    }

    // Delete a Kundli by ID
    public void deleteKundli(UUID kundliId) {
        kundliRepository.deleteById(kundliId);
    }

    // Get all Kundlis
    public List<Kundli> getAllKundlis() {
        return kundliRepository.findAll();
    }
}

File: kundli-astro-frontend/src/services/kundliService.js
import { get, post, put, del } from '../lib/api';
import { KUNDLI_API } from '../lib/constants/apiEndpoints';
import { error } from '../lib/logService';

/**
 * Get a kundli by ID
 * @param {string} id - The kundli ID
 */
export const getKundliById = async (id) => {
  try {
    return await get(KUNDLI_API.GET_KUNDLI_BY_ID(id));
  } catch (err) {
    error(`Failed to get kundli by ID: ${id}`, err);
    throw err;
  }
};

/**
 * Update a kundli by ID
 * @param {string} id - The kundli ID
 * @param {Object} kundliData - The kundli data to update
 */
export const updateKundli = async (id, kundliData) => {
  try {
    return await put(KUNDLI_API.UPDATE_KUNDLI(id), kundliData);
  } catch (err) {
    error(`Failed to update kundli by ID: ${id}`, err);
    throw err;
  }
};

/**
 * Delete a kundli by ID
 * @param {string} id - The kundli ID
 */
export const deleteKundli = async (id) => {
  try {
    return await del(KUNDLI_API.DELETE_KUNDLI(id));
  } catch (err) {
    error(`Failed to delete kundli by ID: ${id}`, err);
    throw err;
  }
};

/**
 * Create a new kundli
 * @param {Object} kundliData - The kundli data to create
 */
export const createKundli = async (kundliData) => {
  try {
    return await post(KUNDLI_API.CREATE_KUNDLI, kundliData);
  } catch (err) {
    error('Failed to create kundli', err);
    throw err;
  }
};

/**
 * Get all kundlis
 */
export const getAllKundlis = async () => {
  try {
    return await get(KUNDLI_API.GET_ALL_KUNDLIS);
  } catch (err) {
    error('Failed to get all kundlis', err);
    throw err;
  }
};

export default {
  getKundliById,
  updateKundli,
  deleteKundli,
  createKundli,
  getAllKundlis,
};
